libraries/chain/controller.cpp

```diff
@@ -1274,6 +1274,59 @@ struct controller_impl {
       apply_handlers[receiver][make_pair(contract,action)] = v;
    }
 
+   void set_trx_expiration(signed_transaction& trx) {
+      if (is_builtin_activated(builtin_protocol_feature_t::no_duplicate_deferred_id)) {
+         trx.expiration       = time_point_sec();
+         trx.ref_block_num    = 0;
+         trx.ref_block_prefix = 0;
+      } else {
+         trx.expiration = time_point_sec{
+            pending_block_time() + fc::microseconds(999'999)}; // Round up to nearest second to avoid appearing expired
+         trx.set_reference_block(chain_head.id());
+      }
+   }
+
+   getpeerkeys_res_t get_top_producer_keys() {
+      try {
+         auto get_getpeerkeys_transaction = [&]() {
+            auto perms = vector<permission_level>{};
+            action act(perms, config::system_account_name, "getpeerkeys"_n, {});
+            signed_transaction trx;
+
+            trx.actions.emplace_back(std::move(act));
+            trx.set_reference_block(chain_head.id());
+            set_trx_expiration(trx);
+            return trx;
+         };
+
+         auto metadata = transaction_metadata::create_no_recover_keys(
+            std::make_shared<packed_transaction>(get_getpeerkeys_transaction()),
+            transaction_metadata::trx_type::read_only);
+
+         const auto& gpo = db.get<global_property_object>();
+
+         auto trace = push_transaction(metadata, fc::time_point::maximum(), fc::microseconds::maximum(),
+                                       gpo.configuration.min_transaction_cpu_usage, true, 0);
```

Comment:
No need to use `gpo.configuration.min_transaction_cpu_usage` for read-only, use `0` for `billed_cpu_time_us`. `explicit_billed_cpu_time` should be `false` not `true`. I think `max_transaction_time` can probably be left as maximum although we might consider using a smaller value. For `block_deadline`, if you are going to call this from `producer_plugin` then I think you should go ahead and use the actual `block_deadline`.

Reply:
https://github.com/AntelopeIO/spring/pull/1316/commits/9c783fd7036de6e1efc44b64fd6d95df9e9a7a9a